text/2020-00-00-error-handling.md

```diff
@@ -0,0 +1,297 @@
+# Error handling
+
+## Summary
+
+Modernise, unify, and simplify error handling in TiKV. I propose using a
+customised or future version of thiserror in all our modules. Migrating TiKV's
+error types will make an excellent 'quest issue' for new contributors.
+
+## Motivation
+
+TiKV takes error handling seriously and has a thorough system of errors.
+Unfortunately, like software tends to do, this system has become large and
+untidy as it has grown. TiKV has more than 30 different error types using two
+different error libraries (as well as `std::Error`).
+
+TiKV has some very large error types which are inefficient to pass by value. So
+in some places (e.g.,
+[src/storage](https://github.com/tikv/tikv/blob/master/src/storage/errors.rs))
+we use a pattern of having `Error` and `ErrorInner` types and boxing the inner
```

Comment:
I think `ErrorInner` exists because we want to achieve minimal copy cost. It's not caused by `quick_error` library. Hence this should be able to be solved without switching the error library, right?

Reply:
I think one of the goal of this RFC is to be modern (i.e. utilize the new `std::Error`).

Reply:
`quick_error` has been using `std::error::Error` from the beginning.

Reply:
This RFC has a number of goals that are mostly unrelated, other than that they are easier to implement at once. I have looked at our error handling in a holistic way to find ways to improve. The `ErrorInner` stuff is mostly orthogonal to the choice of library, but using thiserror makes it easier to implement the extensions than using quick_error. I think thiserror is a better library because it does not reimplement anything from std::Error and its syntax is more succinct and idiomatic.